### PR TITLE
feat: added external partitioned Delta table

### LINKED ISSUE #122: Create built-in Partitioned Delta table
We have a bunch of built-in tables that are great for localhost experimentation:

Screenshot 2024-07-01 at 8 22 28 AM

It'd be nice to add a partitioned Parquet table as well.  Perhaps the table can have `first_name`, `age` and `country` columns and be partitioned by `country`.

The example would also serve as a good example of [how Unity Catalog supports hive-style partitioning](https://github.com/unitycatalog/unitycatalog/issues/106).

### PR DESCRIPTION
**Pull request by SemyonSinchenko**
**PR Checklist**

- [x] A description of the changes is added to the description of this PR.
- [x] If there is a related issue, make sure it is linked to this PR. (close #122)
- [x] If you've fixed a bug or added code that should be tested, add tests!
- [x] If you've added or modified a feature, documentation in `docs` is updated

**Description of changes**

I added a small partitioned Delta table:

- Table was created with delta-rs
- I modified manually db-file to remove my local path (it starts from `etc/...` now)
- I removed from _delta_log a location (from commitInfo) that contains my local path

(with assumption that deltalake, faker, pandas are installed)

```python
pdf = pd.DataFrame([{"first_name": fake.name(), "age": random.randint(20, 75), "country": countries[random.randint(0, 2)]} for _ in range(30)])
write_deltalake("~/github/unitycatalog/etc/data/external/unity/default/tables/user_countries", data=pdf, partition_by="country")
```

### CODE REVIEW

**Comment by SemyonSinchenko** [comment]
@MrPowers Take a look, please!

**Comment by SemyonSinchenko** [comment]
image

**Comment by SemyonSinchenko** [comment]
I think that I should wait until #156 is merged, shouldn't I?

**Comment by SemyonSinchenko** [comment]
@ravivj-db Hello! May you take a look, please? I updated `PopulateTestDatabase.java` by adding an information about external partitioned Delta table. There are not unit-tests for that, but I tested it locally by deleting the db-file and run `sbt populateTestDb`.

image

File: server/src/test/java/io/unitycatalog/server/utils/PopulateTestDatabase.java
Comment (ravivj-db) [inline server/src/test/java/io/unitycatalog/server/utils/PopulateTestDatabase.java:270]: For creating external table, you can just use tableOperations as it's already supported.

Comment (ravivj-db) [reply server/src/test/java/io/unitycatalog/server/utils/PopulateTestDatabase.java:270]: Also, please put a more detailed comment.

Comment (SemyonSinchenko) [reply server/src/test/java/io/unitycatalog/server/utils/PopulateTestDatabase.java]: > just use tableOperations

May you explain it little more, please? I see that both implementation of `TableOperations` required running UC Server. But in the `PopulateTestDatabase.java` there is no db and running server. Thanks in advance!

P.S> I put a more detailed comment, please take a look is a new one enough? Tnx in advance!

Comment (ravivj-db) [reply server/src/test/java/io/unitycatalog/server/utils/PopulateTestDatabase.java]: Sorry Since this is PopulateTestDatabase Class, you can create/get an instance of TableRepository (Similar Catalog and Schema Repository) and use that to create an external table (using a CreateTable Request). and Maybe simplify the creation of previously added external tables as well.

Comment (ravivj-db) [reply server/src/test/java/io/unitycatalog/server/utils/PopulateTestDatabase.java]: When you instantiate the TableRepository it will connect to the default db on its own (via HibernateUtil).

Comment (SemyonSinchenko) [reply server/src/test/java/io/unitycatalog/server/utils/PopulateTestDatabase.java]: Ok, let me try to do it!

Comment (SemyonSinchenko) [reply server/src/test/java/io/unitycatalog/server/utils/PopulateTestDatabase.java]: @ravivj-db I rewrite one of tables to using `TableRepository`. Please, take a look and say what do you think about it, before I changed previously added tables, catalogs, schemas, etc. Because from my point of view the new code does not like like simplification... Tnx in advance!

File: server/src/test/java/io/unitycatalog/server/utils/PopulateTestDatabase.java
Comment (ravivj-db) [inline server/src/test/java/io/unitycatalog/server/utils/PopulateTestDatabase.java:311]: Setting these generic values might pose problems in a third party client which is trying to interpret these columns (We observed the same with duckdb). Make sure you set these values keeping in mind the consumers. This goes for both `typeText` and `typeJson`.

File: server/src/test/java/io/unitycatalog/server/utils/PopulateTestDatabase.java
Comment (ravivj-db) [inline server/src/test/java/io/unitycatalog/server/utils/PopulateTestDatabase.java:295]: use constants `catalogName` and `schemaName` here.

File: server/src/test/java/io/unitycatalog/server/utils/PopulateTestDatabase.java
Comment (ravivj-db) [inline server/src/test/java/io/unitycatalog/server/utils/PopulateTestDatabase.java:324]: why is partition index only being set for `country` column.

File: server/src/test/java/io/unitycatalog/server/utils/PopulateTestDatabase.java
Comment (ravivj-db) [reply server/src/test/java/io/unitycatalog/server/utils/PopulateTestDatabase.java]: Have reviewed, besides minor comments overall looks ok. Also, let's just merge this one for now. We can change the way previously added EXTERNAL tables are being created in a separate PR.

File: server/src/test/java/io/unitycatalog/server/utils/PopulateTestDatabase.java
Comment (SemyonSinchenko) [reply server/src/test/java/io/unitycatalog/server/utils/PopulateTestDatabase.java:324]: As I understood [from the answer](https://github.com/unitycatalog/unitycatalog/discussions/217#discussioncomment-10031239), it is how partitioning information is handled by UC. Like it is the first partition column and if I would like to add the nested partitioning, the next column will have an index `1`.

File: server/src/test/java/io/unitycatalog/server/utils/PopulateTestDatabase.java
Comment (SemyonSinchenko) [reply server/src/test/java/io/unitycatalog/server/utils/PopulateTestDatabase.java:311]: Without setting these values (`typeName`, `typeText`) I was getting an NPE from [this code block](https://github.com/unitycatalog/unitycatalog/blob/209895f0409a8bf87f6a7fa0de30d1868d6d6727/server/src/main/java/io/unitycatalog/server/persist/TableRepository.java#L109)...

Comment (ravivj-db) [reply server/src/test/java/io/unitycatalog/server/utils/PopulateTestDatabase.java:311]: I just meant, make sure your final client understands these values, whatever being set here.

**Review by ravivj-db — COMMENTED** [review]
Any reason why the db file changed in this commit ? My guess is create/update Dates.

File: server/src/test/java/io/unitycatalog/server/utils/PopulateTestDatabase.java
Comment (ravivj-db) [reply server/src/test/java/io/unitycatalog/server/utils/PopulateTestDatabase.java:324]: @tdas  can you please review this one ?

**Review by ravivj-db — APPROVED** [review]
@tdas  feel free to merge this one, once you have confirmed the setting of the partition index in the column info.

File: server/src/test/java/io/unitycatalog/server/utils/PopulateTestDatabase.java
Comment (SemyonSinchenko) [reply server/src/test/java/io/unitycatalog/server/utils/PopulateTestDatabase.java:313]: @ravivj-db Thanks for mentioning it! I made it work with DuckDB by modifying how I pass the JSON value and also by adding required by DuckDB precision and scale.

**Comment by SemyonSinchenko** [comment]
> Any reason why the db file changed in this commit ? My guess is create/update Dates.

Yes, create/update Dates and also I updated precision, scale and JSON types representation for the added table.

**Comment by ravivj-db** [comment]
@SemyonSinchenko  there's one more ask. Thanks for being patient so far :). Can you add quickstart tests to tests/test_tutorial.py for reading metadata and contents of the table (using get and read)

**Comment by ravivj-db** [comment]
This might require adding read functionality for Parquet Data Format in TablesCLI. If that's too much you can just add a test for getTable Metadata.

**Comment by SemyonSinchenko** [comment]
> @SemyonSinchenko there's one more ask. Thanks for being patient so far :). Can you add quickstart tests to tests/test_tutorial.py for reading metadata and contents of the table (using get and read)

Done: [commit](https://github.com/unitycatalog/unitycatalog/pull/151/commits/bbc1f0a1bd37585f2103f89683eb21b5e5a9006c)

**Comment by ravivj-db** [comment]
One of the tests is failing, can you please take a look ?

**Comment by ravivj-db** [comment]
@SemyonSinchenko just read your message, I understand the complication with using TableRepository, and I agree with you that we can switch back to using Hibernate/TableDAO (I guess that's why we used them in the first place :)). I am really sorry about this back and forth.

**Comment by SemyonSinchenko** [comment]
> @SemyonSinchenko just read your message, I understand the complication with using TableRepository, and I agree with you that we can switch back to using Hibernate/TableDAO (I guess that's why we used them in the first place :)). I am really sorry about this back and forth.

@ravivj-db Don't worry :)
I switched back to DAOs, do you want to take a final look again?

**Review by ravivj-db — APPROVED** [review]
LGTM

**Review by tdas — APPROVED** [review]

**Comment by tdas** [comment]
This is great! Thank you @SemyonSinchenko for adding this. huge help!

**Pull request merged**
